compiler/src/dotty/tools/dotc/transform/localopt/RemoveUnnecessaryNullChecks.scala

```diff
@@ -6,104 +6,118 @@ import core.Contexts.Context
-  }
+       
+        visitBranches(cond, th, el)
+
+      case _ =>
+    } 
+
+  // return nullchecked symbol in expr (if any)
+  private def nullChecksInExpr(expr: Tree)(implicit ctx: Context): Option[Symbol] =
+    expr match {
+      case t @ Apply(Select(id: Ident, op), List(rhs)) if (t.symbol == defn.Object_eq || op == nme.EQ) && !isVar(id.symbol) && isNull(rhs) => Some(id.symbol)
+      case t @ Apply(Select(lhs, op), List(id: Ident)) if (t.symbol == defn.Object_eq || op == nme.EQ) && !isVar(id.symbol) && isNull(lhs) => Some(id.symbol)
+      case _ => None
+    }
 
+  // return all sub-trees of tree containing a nullcheck for symbol
+  private def findNullChecksFor(symbol: Symbol, tree: Tree)(implicit ctx: Context) = 
+    tree.filterSubTrees(nullChecksInExpr(_) match { case Some(s) if (s == symbol) => true; case _ => false })
+
+  // fill replace using findNullChecksFor
+  private def flagAsNull(symbol: Symbol, tree: Tree)(implicit ctx: Context) = replace ++= findNullChecksFor(symbol, tree).map((_, Literal(Constant(true))))
+  private def flagAsNotNull(symbol: Symbol, tree: Tree)(implicit ctx: Context) = replace ++= findNullChecksFor(symbol, tree).map((_, Literal(Constant(false))))
+
+  private def isNotNull(tree: Tree)(implicit ctx: Context): Boolean =
+    tree match {
+      case Block(_, expr) => isNotNull(expr)
+      case If(_, th, el) => isNotNull(th) && isNotNull(el)
+      case t: Typed => isNotNull(t.expr)
+      case t: Literal => t.const.tag != NullTag
+      case _: This => true
+      case _: New => true
+      case t: Apply if t.symbol.isPrimaryConstructor => true
+      case Apply(Select(New(_), _), _) => true
+      case t => t.tpe.isNotNull
+    }
 
-  def transformer(implicit ctx: Context): Tree => Tree = {
-    def isNullLiteral(tree: Tree) = tree match {
-      case literal: Literal =>
-        literal.const.tag == NullTag
+  private def isNull(tree: Tree)(implicit ctx: Context): Boolean = 
+    tree match {
+      case Block(_, expr) => isNull(expr)
+      case If(_, th, el) => isNull(th) && isNull(el)
+      case t: Typed => isNull(t.expr)
+      case t: Literal => t.const.tag == NullTag
+      case EmptyTree => true
       case _ => false
     }
-    val transformation: Tree => Tree = {
-      case check @ Apply(Select(lhs, _), List(rhs)) =>
-        val sym = check.symbol
-        val eqOrNe  = sym == defn.Object_eq || sym == defn.Object_ne
-        val nullLhs = isNullLiteral(lhs) && isGood(rhs.symbol)
-        val nullRhs = isNullLiteral(rhs) && isGood(lhs.symbol)
-
-        if (eqOrNe && (nullLhs || nullRhs)) {
-          def block(b: Boolean) = Block(List(lhs, rhs), Literal(Constant(b)))
-          if (sym == defn.Object_eq) block(false)
-          else if (sym == defn.Object_ne) block(true)
-          else check
-        } else check
-
-      case t => t
-    }
-    transformation
-  }
+
+  private def isVar(s: Symbol)(implicit ctx: Context) = s.is(Mutable, Lazy) && !s.is(Method) && !s.owner.isClass
```

Comment:
Why do you need anything more than `.is(Mutable)` here? Also I think this version of `is` was intended to always be used with named second argument: `s.is(Mutable, butNot = Lazy)`.